R/data_rename.R

```diff
@@ -96,28 +97,34 @@
 #'
 #' @export
 data_rename <- function(data,
-                        pattern = NULL,
+                        select = NULL,
                         replacement = NULL,
                         safe = TRUE,
                         verbose = TRUE,
+                        pattern = NULL,
                         ...) {
+  if (!is.null(pattern)) {
+    .is_deprecated("pattern", "select")
+    select <- pattern
+  }
+
   # change all names if no pattern specified
-  if (is.null(pattern)) {
-    pattern <- names(data)
+  if (is.null(select)) {
+    select <- names(data)
   }
```

Comment:
I think at this point we can directly use the code from other functions, like `data_select()`, which calls `.select_nse()`. line 112 could be:

```r
  select <- .select_nse(
    select,
    data,
    exclude,
    ignore_case = ignore_case,
    regex = regex,
    allow_rename = TRUE,
    verbose = FALSE
  )
```

Note the `allow_rename = TRUE`, which still allows named vectors for renaming.

Reply:
Actually this would change several behavior / messages. For instance:
```r
data_rename(head(iris), select = c(test = "Species", "Sepal.Length"))
```
errored before because there is a missing replacement, but now `.select_nse()` returns a named vector with `Sepal.Length` as value and name, which means the rest of the function doesn't error anymore.

Also, the argument `safe` would be useless because `.select_nse()` returns an empty character vector if some column names are unknown, and therefore this argument is not reached anymore. Just want to discuss about those before going further with the changes.

Reply:
Regarding `safe`: I'd say we can remove this argument, and set `verbose = TRUE` when calling `.select_nse()`. First, this will warn users about non-existent columns (`.warn_not_found()` is called) and second, we error later if number of variables to rename, and number of new names in `rename` do not match.

Regarding not all elements named: That's a good point. We could argue that we trust users that they know what they're doing. Or we could even extend into something like:

```r
data_rename(mtcars, select = c(test = "mpg", "gear", "hp), rename = "new_{n}")
```

which would use names (if not identical) as new value to rename, and for those where names = values (`gear` and `hp` here), we use the values specified in `rename` (though, if we allow to mix rename-values mixed with/without tokens, it's much more complicated - maybe we just say if users know what they do and let `data_rename(head(iris), select = c(test = "Species", "Sepal.Length"))` just happen...)

Reply:
I think overall, the benefits of having the classic `select` functionality outweighs the new "limitations" we get.

Reply:
> First, this will warn users about non-existent columns (.warn_not_found() is called) and second, we error later if number of variables to rename, and number of new names in rename do not match.

This was actually not the case. We only throw a message when `length(replacement) != length(select)`, which I think is not the right thing to do. I've changed it to an error, otherwise there's no replacement for `safe = TRUE` (the default).

> We could argue that we trust users that they know what they're doing. [...] `data_rename(head(iris), select = c(test = "Species", "Sepal.Length"))`

Actually I think using column indices as default column names if no replacement is provided is a weird behavior. Isn't a single number a "syntactically invalid name" as in #567? Do you have any usecase with this?

Reply:
But why column indices? I think `select = c(test = "Species", "Sepal.Length")` returns a fully named vector, meaning that variables with no new names are "ignored":

``` r
datawizard::data_select(head(iris), select = c(test = "Species", "Sepal.Length"))
#>     test Sepal.Length
#> 1 setosa          5.1
#> 2 setosa          4.9
#> 3 setosa          4.7
#> 4 setosa          4.6
#> 5 setosa          5.0
#> 6 setosa          5.4
```

<sup>Created on 2024-11-30 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

Reply:
Right, I got confused with the behavior for unnamed vectors (e.g. just specifying `select`) where column indices are used:
``` r
datawizard::data_rename(head(iris), select = c("Species", "Sepal.Length"))
#>     2 Sepal.Width Petal.Length Petal.Width      1
#> 1 5.1         3.5          1.4         0.2 setosa
#> 2 4.9         3.0          1.4         0.2 setosa
#> 3 4.7         3.2          1.3         0.2 setosa
#> 4 4.6         3.1          1.5         0.2 setosa
#> 5 5.0         3.6          1.4         0.2 setosa
#> 6 5.4         3.9          1.7         0.4 setosa
```

So two things:
* I think we should keep the behavior in `main` and error if the vector is named but some elements don't have names
* using column indices when no replacement is provided seems like a bad practice and we could error because `length(select) > length(replacement)`

Reply:
Yes, agreed. Only if `replacement` is `NULL` and no named vector in `select`, then sequential numbering is intended.

Reply:
Ah sorry I should have been clearer. For the second point, I meant that using column indices as replacement is a bad practice in all cases. For example, IMO this should error:
```r
datawizard::data_rename(head(iris), select = c("Species", "Sepal.Length"))
```
should error.

Reply:
Yes, yes, I understood. But I think this was an intentional behavior, if I recall right:

![Screenshot_20241130-154926](https://github.com/user-attachments/assets/2f978fa9-d5b4-46df-a4c5-bcb1828ccc64)

I think @DominiqueMakowski wanted that feature?

Reply:
> For a named vector, .select_nse() always returns a fully named vector. For non-named elements, the element value is used as name. So if we want to check whether some elements don't have names, we would do this separately, right?

This could be done before `.select_nse()`, like this:
```r
if (!is.null(names(select)) && any(names(select) == "")) {
    insight::format_error("When `select` is a named vector, all elements must be named.")
}
```
I think mixing behaviors, e.g. allowing this `data_rename(x, c("foo" = "Species", "Sepal.Length"), "new_{col}")` would lead to confusing code, which is why I'd rather detect early when some elements of `select` are named and others aren't.

> we would need to modify .warn_not_found() in a way it could either warn or error, if we want to error on not-found variables. 

That's possible to do, but it would complexify a bit more the internals of `.select_nse()`. Maybe we could just catch the warning in `data_rename()` and convert it to error? Or just remove `safe`? (although I can see when this is useful, even more since we have the glue-style replacement).

> I think @DominiqueMakowski wanted that feature?

I'd be curious to know in which context renaming columns "1", "2", etc. is useful.

Reply:
> This could be done before .select_nse(), like this:

That's ok for me, but doesn't this evaluate the `select` argument? What if `select` is not a character vector?

> Or just remove safe?

I'd say we remove `safe` and warn/error on every unclear case.

> That's possible to do, but it would complexify a bit more the internals of .select_nse(). Maybe we could just catch the warning in data_rename() and convert it to error? 

I think it's less complex than catching a warning and throwing an error. We just need to pass an argument like `action` to `.warn_not_found()`, and there we have `format_alert(..., type = action)`. Much less code, or not?

> I'd be curious to know in which context renaming columns "1", "2", etc. is useful.

Not sure - @DominiqueMakowski ?

Reply:
> That's ok for me, but doesn't this evaluate the select argument? What if select is not a character vector?

Right, I've done another approach in [`bd0be81` (#568)](https://github.com/easystats/datawizard/pull/568/commits/bd0be816d67090d372c79f35bfb5234a5fa06b66)

> I think it's less complex than catching a warning and throwing an error. We just need to pass an argument like action to .warn_not_found(), and there we have format_alert(..., type = action). Much less code, or not?

It adds an arg in all internal functions of NSE, see [`f2a56af` (#568)](https://github.com/easystats/datawizard/pull/568/commits/f2a56af1b29b2aa967924b389b7557cc405be4cb)

I had removed the behavior of using column indices for names, but I'll fix the remaining tests and examples once I know if we keep this behavior or not.

Reply:
> It adds an arg in all internal functions of NSE

Right, it's indeed in more places than I thought...

> Right, I've done another approach in bd0be81 (#568)

Looks good, but could it be shorter? Instead of the for-loop, compare all names against elements:

```r
if (!is.null(names(select)) && any(names(select) == select)) {
...
}
```

Reply:
Done in [`62631e6` (#568)](https://github.com/easystats/datawizard/pull/568/commits/62631e6ec9037fdae32efe58c5b4975ca76add68)